executor/internal/pdhelper/pd.go

```diff
@@ -0,0 +1,123 @@
+// limitations under the License.
+
+package pdhelper
+
+import (
+	"context"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/jellydator/ttlcache/v3"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/store/helper"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/sqlexec"
+)
+
+// GlobalPDHelper is the global variable for PDHelper.
+var GlobalPDHelper *PDHelper
+var globalPDHelperOnce sync.Once
+
+func init() {
+	GlobalPDHelper = newPDHelper()
+}
+
+// PDHelper is used to get some information from PD.
+type PDHelper struct {
+	wg                                       util.WaitGroupWrapper
+	cacheForApproximateTableCountFromStorage *ttlcache.Cache[string, float64]
+}
+
+func newPDHelper() *PDHelper {
+	cache := ttlcache.New[string, float64](
+		ttlcache.WithTTL[string, float64](30*time.Second),
+		ttlcache.WithCapacity[string, float64](1024*1024),
+	)
+	return &PDHelper{
+		cacheForApproximateTableCountFromStorage: cache,
+	}
+}
+
+// Start starts the PDHelper.
+func (p *PDHelper) Start() {
+	globalPDHelperOnce.Do(func() {
+		p.wg.Run(p.cacheForApproximateTableCountFromStorage.Start)
+	})
+}
+
+// Stop stops the PDHelper.
+func (p *PDHelper) Stop() {
+	p.cacheForApproximateTableCountFromStorage.Stop()
+	p.wg.Wait()
+}
+
+func approximateTableCountKey(tid int64, dbName, tableName, partitionName string) string {
+	return strings.Join([]string{strconv.FormatInt(tid, 10), dbName, tableName, partitionName}, "_")
+}
+
+// GetApproximateTableCountFromStorage gets the approximate count of the table.
+func (p *PDHelper) GetApproximateTableCountFromStorage(sctx sessionctx.Context, tid int64, dbName, tableName, partitionName string) (float64, bool) {
+	key := approximateTableCountKey(tid, dbName, tableName, partitionName)
+	if item := p.cacheForApproximateTableCountFromStorage.Get(key); item != nil {
+		return item.Value(), true
+	}
+	tikvStore, ok := sctx.GetStore().(helper.Storage)
+	if !ok {
+		return 0, false
+	}
+	regionStats := &helper.PDRegionStats{}
+	pdHelper := helper.NewHelper(tikvStore)
+	err := pdHelper.GetPDRegionStats(tid, regionStats, true)
+	failpoint.Inject("calcSampleRateByStorageCount", func() {
+		// Force the TiDB thinking that there's PD and the count of region is small.
+		err = nil
+		regionStats.Count = 1
+		// Set a very large approximate count.
+		regionStats.StorageKeys = 1000000
+	})
+	if err != nil {
+		return 0, false
+	}
+	// If this table is not small, we directly use the count from PD,
+	// since for a small table, it's possible that it's data is in the same region with part of another large table.
+	// Thus, we use the number of the regions of the table's table KV to decide whether the table is small.
+	if regionStats.Count > 2 {
+		result := float64(regionStats.StorageKeys)
+		p.cacheForApproximateTableCountFromStorage.Set(key, result, ttlcache.DefaultTTL)
```

Comment:
What will happen when capacity is full?

Reply:
No, The Least Recently Used one will be deleted.